### PR TITLE
Facilitate User Taxonomies

### PR DESCRIPTION
**Pull request by KTS915**
It is already possible in ClassicPress to create taxonomies for users. (To do so, simply create a taxonomy in the usual way, but register the taxonomy with `user` as the second argument:
```
register_taxonomy( 'name-of-taxonomy, 'user', $args );
```
The problem is that there is no user interface built-in that takes advantage of the creation of user taxonomies. This PR fills that gap by creating a sub-menu item under Users in the admin pages for each user taxonomy created. This sub-menu item then links to pages where new taxonomy terms may be added, edited, and deleted in the same way — and with an identical-looking UI — as with posts.

This functionality has been requested on WP's Trac system many times. Those requests all got funneled to [Trac Ticket 31383](https://core.trac.wordpress.org/ticket/31383), where some code was proposed. This PR includes that code but goes further in the following respects:

- URLs refer to users instead of to posts
- A new column is added to the `users.php` page for each user taxonomy created, and this column is populated by the name(s) of the taxonomy term(s) to which the user belongs
- These terms are hyperlinked so that, if clicked, they enable the filtering of the list of users so that only those in the clicked taxonomy are shown
- Checkboxes are added to the `user-edit.php` page, so that administrators and/or users themselves may add a user to one or more user taxonomy terms. (Note that the Trac ticket says that the number of users will not be automatically be updated because the core function for that applies only to posts, but that is no longer true, so no new code is required for that purpose.)

Filters are provided throughout, so that the default settings may be easily changed. But it is particularly important to note that, unless a user taxonomy is registered on a specific site, nothing about that site's appearance will change. This means that testing will require the creation of a user taxonomy using the code shown above.

## Motivation and context
I have several sites that have user taxonomies, and it's awkward to have to write somewhat convoluted code in a plugin to make this functionality as seamless as it should be. This PR makes having a user taxonomy as painless an experience as having a custom taxonomy for posts.

Note that, once user taxonomies are used, it may well generate new momentum for the aspiration to have an easy way to be able to relate different objects to each other beyond posts and taxonomies. I have some code that does precisely that which I use on a current production site, but that can wait for a later PR.

## How has this been tested?
The specific code in this PR has been tested on my localhost test site. The code on which it is based, however, has been in use on production sites for around two years.

## Screenshots using "Law Schools" as a user taxonomy
Screenshot at 2024-04-11 15-39-16
Screenshot at 2024-04-11 15-39-54
Screenshot at 2024-04-11 15-40-31

## Types of changes
- New feature, but not a breaking change

### CODE REVIEW

File: tests/phpunit/tests/taxonomy.php [hunk 1]
```diff
@@ -33,6 +33,16 @@ public function test_get_post_taxonomy() {
 		}
 	}
 
+	function test_get_user_taxonomy() {
```
Comment (xxsimoxx) [inline tests/phpunit/tests/taxonomy.php:36]: This test seems unnecessary as 'user' has not built in taxonomies as 'post'.

File: src/wp-includes/taxonomy.php [hunk 2]
```diff
@@ -684,7 +684,7 @@ function register_taxonomy_for_object_type( $taxonomy, $object_type ) {
 		return false;
 	}
 
-	if ( ! get_post_type_object( $object_type ) ) {
+	if ( 'user' !== $object_type && ! get_post_type_object( $object_type ) ) {
```
Comment (xxsimoxx) [inline src/wp-includes/taxonomy.php:687]: `unregister_taxonomy_for_object_type` seems to need the same change (this fixes two failing tests).

File: tests/phpunit/tests/taxonomy.php (hunk 1 above)
Comment (KTS915) [reply tests/phpunit/tests/taxonomy.php]: I wondered. On the original Trac ticket, someone asked whether this was right, but it never got answered.

**Comment by KTS915** [comment]
I'm now seeing this message on admin pages:
```
preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated
wp-admin/menu-header.php:63
```

**Comment by mattyrob** [comment]
I've done some testing on this and it works as described. I have checked that users below `Administrator` level won't be able to access this - I presume that is as expected and hoped. I can't really comment on usability as this is not a feature (custom post types taxonomies etc) that I use.

**Comment by KTS915** [comment]
Thanks for testing, @mattyrob! The permissions work as intended by default, but there are plenty of filters that enable them to be changed if desired.

**Review by xxsimoxx — APPROVED** [review]

**Comment by KTS915** [comment]
@tradesouthwest I am not sure why you commented here, since this PR has already been merged, but you seem to have misunderstood what this PR addresses. It's not about user capabilities at all but user _taxonomies_.

**Pull request merged**